### PR TITLE
Dev metaT notebook

### PR DESCRIPTION
**Pull request by RodolfoSalido**
added a metatranscriptomics specific notebook for NPH.

### CODE REVIEW

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:116]: Does this comment mean the user is supposed to input something here?  If so, I think it needs more elaboration!  But if it is a left-over, let's take it out.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:171]: Why is this defined here but not used in this block?

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:207]: Why is this re-defined here but also not used in this block?

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:225]: replace "concentratin" with "concentration"

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:226]: I vote to define well_col in this code block, where it is first actually used.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:277]: A comment explaining what the hard-coded threshold=40 means would be helpful for maintenance in the future.

Also, I see that it is used in two places in this block.  If these two uses should always be the same, then it might be a good idea to put the threshold value in a variable and use the variable in both places, to make the value easier to change if needed.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:308]: this is checking if `sample_dna_concs_fp` already exists, not if `dnase_dna_concs_fp` already exists; I think it should be checking the latter.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:341]: Again, I don't think this should be (re)-defined here if it is not being used in this cell.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:377]: Same comment as above.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:396]: Obviously, one can get away with not defining well_col in this cell because it has been defined above.  However, I vote for explicitly redefining it here ... it makes debugging difficult when one has to search through the whole notebook to find the last most recent re-definition of a variable.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:637]: Daniel deeply wants read_csv() calls to use dtype=str so that they read everything in as a string rather than allowing automatic type selection (which profoundly screws up things like tube codes).  However, my guess is that you might want the index column to be read in as an int.  Would you consider reading everything in as a str and then explicitly converting the columns you need to be not strings into the type you want them to be?

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1062]: It would be helpful to add a comment explaining where the max_vol_per_well=30000 hardcoding came from.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1119]: remove period at end of bulleted phrase

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1129]: replace "requires" with "allows"

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1157]: Fix formatting: the `**` for bolding is opened right before `Project Name` but not closed, thus causing the `**` later for `Sample_Project` to read wrong.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1268]: See above re Daniel's wish for dtype specification for read_csv() ...

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1351]: It would be useful to have a comment describing the source of the dynamic_range=20 hard-coding, and info on whether the user should ever be changing this value.

File: metapool/metapool.py
Comment (AmandaBirmingham) [inline metapool/metapool.py:374]: It would be helpful to have a comment about where the max_conc=250 hardcoding came from, and under what circumstances it might change.

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (AmandaBirmingham) [inline notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:1385]: Why did this change?

**Review by AmandaBirmingham — CHANGES_REQUESTED** [review]
Request a few small typo fixes and some clarifications of default values.  Looks good!

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (charles-cowart) [inline notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1]: Having trouble with running the new metatranscriptomics script, although the issues appear to occur with the metagenomics script as well. I'll investigate further but in short this notebook should function without errors before merging.

**Review by charles-cowart — CHANGES_REQUESTED** [review]
A substantial PR - I have some questions and comments that should be reviewed before approving. I will continue to look into the issue with running the notebooks.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:36]: will fix this

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:116]: it is somewhat of a vestige for a functionality that is not being used routinely. Maybe @charles-cowart or @antgonza can comment? I believe the qiita.oauth.cfg file was used for authenticating a user functionality that would allow the notebook to check sample names against the metadata associated with the study.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:171]: I can clean this up, thanks for pointing this out.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:207]: I'll correct this.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:226]: sounds good. Will do.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:225]: will fix

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:277]: sounds good. Will include the comment explaining the threshold value and potentially define it in a variable

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:308]: you are correct, thanks for catching this. Will fix it

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:792]: sounds good, will do.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:980]: N = total number of samples in the prep. I can add a clarification.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1014]: When I wrote the function 100 µL seemed like a good default value. After putting this function in practice, we figured that pooling 190µL in total would result in ~500nL per sample for a full 384 well plate, so we hardcoded the value in the notebook to 190µL, since that is the value that wetlab technicians were using most often. I can change the default value to 190µL.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1062]: will add an explanation.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1351]: this is actually a parameter that the technician has to play with until they find the optimal value. I will add an explanation.

File: metapool/metapool.py
Comment (RodolfoSalido) [reply metapool/metapool.py:1223]: Ok. I will test this regex pattern instead.

File: notebooks/test_data/Quant/MiniPico/2024_NPH_Plates_3-5_gDNA_quant.txt
Comment (RodolfoSalido) [reply notebooks/test_data/Quant/MiniPico/2024_NPH_Plates_3-5_gDNA_quant.txt:1]: I think these (the utf-16 files) are all direct output files from the spectramax plate reader that we use for nucleic acid quantification. I will double check.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1]: what is the issue you are encountering?

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (charles-cowart) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:637]: It is interesting that historically we've wrapped sample-sheets in objects that incorporate custom type conversion and validation, but we don't do it for csvs of other types. At the least we should migrate people towards using a wrapper function that loads csvs into data frames as all dtype=str. this would make periodically searching and replacing new instances of read_csv() a non-issue. custom openers for different csvs with different headers might be a possible solution as well.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (charles-cowart) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1]: Fresh VM + installation is giving me issues with importing. Will revisit shortly.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (RodolfoSalido) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:637]: I am open to making the change, but do we want to do that during this pull request?

Comment (AmandaBirmingham) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:637]: Yeah, I honestly don't know. What I learned from making changes to the matrix notebook is that it is not safe to just add a `dtype=str` to an existing `read_csv`, because in some places, later functionality depends on some columns being, e.g., numbers.  You have to actually know what is being read in and how it is being used later before it is safe to make this change.  I was thinking Rodolfo might have the knowledge necessary to make the change safely here, but if not, I don't want to slow things down for it.

I do not think I am up for trying to add a wrapper function in this PR since we need it to be merged as soon as we can.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1268]: Also see my above comment re Daniel's suggestion being potentially tricky, and not worth doing unless you are very comfortable with how the data being read in is being used :)

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (AmandaBirmingham) [inline notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:40]: replace "assocaited" with "associated"

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (AmandaBirmingham) [inline notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:44]: Replace "Next you'll" with "You'll then"

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (AmandaBirmingham) [inline notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:46]: Just checking: this is a *different* quantification file than the one referenced above, right?  I see **quantification file** twice close together, and my immediate thought is that both those refer to the same thing, but I think they don't ...

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (AmandaBirmingham) [inline notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:395]: Question: does this mean `Well` is a 96-well plate position and `Library Well` is a 384-well plate position?

File: notebooks/metagenomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [inline notebooks/metagenomics_pipeline_iSeqnorm.ipynb:996]: Very helpful, thank you!

**Comment by AmandaBirmingham** [comment]
Just a tiny typo; other than that, I'm good with this PR.  Of course, @charles-cowart has final say since he's checking investigating possible run issues.

**Review by AmandaBirmingham — APPROVED** [review]

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (RodolfoSalido) [reply notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:46]: They are different quantifications. The first quantification file quantifies gDNA concentration, while the second one quantifies sequencing library concentration. I can make a clarification.

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (RodolfoSalido) [reply notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:395]: not exactly. The 'Well' column in the plate_df tracks the well location, in 384-well format, where the gDNA of a sample is contained while the 'Library Well' tracks the well location for the sequencing library of a sample. Most of the time the 'Well' and the 'Library Well' will match, but that is not always the case. For instance when doing plate replication, a sample from 1 'Well' can be replicated and placed in up to 4 different 'Library Wells'.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (charles-cowart) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:116]: @RodolfoSalido That is correct! It is for use with read_plate_map_csv() that will optionally connect to Qiita and compare sample-names in a project with what's in the plate-map file. It's still an option that's available in the sample-sheet creation notebook (and perhaps more?).

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (charles-cowart) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:641]: I agree it doesn't sound like the time and place for a wrapper. This comment doesn't appear to be outdated. Can I assume it was addressed?

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (AmandaBirmingham) [reply notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:395]: thanks!

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:641]: I can live with leaving this out for now.

File: notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb
Comment (charles-cowart) [reply notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb:1]: Resolved the merge conflicts with the current main branch and re-linted the code. Did a fresh pull of the branch and installed it into a fresh environment. both sample-sheets ran through all steps without error. We can call this one closed.

File: metapool/metapool.py
Comment (charles-cowart) [reply metapool/metapool.py:1313]: Okay. this wasn't changed but I don't think it's anything to hold back this PR for. At worst I can run some tests and change it later.

File: metapool/metapool.py
Comment (charles-cowart) [reply metapool/metapool.py:1318]: I'll assume so.

**Review by charles-cowart — APPROVED** [review]

File: metapool/metapool.py
Comment (RodolfoSalido) [reply metapool/metapool.py:1313]: I randomly remembered this suggestion today. Maybe I should raise an issue ? I agree that the potential for a wrong regex match might surface in the future.

**Comment by charles-cowart** [comment]
Sure, that sounds good, thank you!

**Pull request merged**
